### DISCUSSION TITLE
Is there a way to know if a font is Truetype or OpenType?

### DISCUSSION THREAD

**Discussion by moi15moi** [fonttools/fonttools#2619] [Q&A] [answered]
Is there a way to know if a font is Truetype or OpenType via fonttools?

**Comment by justvanrossum** [comment, marked as answer]
```python
from fontTools.ttLib import TTFont

f = TTFont("path/to/font")
if "glyf" in f:
    print("font is TTF")
else:
    print("font is OTF")
```

**Reply by moi15moi** [reply]
Thanks :)

**Reply by moi15moi** [reply]
> ```python
> from fontTools.ttLib import TTFont
> 
> f = TTFont("path/to/font")
> if "glyf" in f:
>     print("font is TTF")
> else:
>     print("font is OTF")
> ```

This font return me "font is TTF", but it is an otf font: https://mega.nz/file/P9EwHZ4K#dqVUewtPbNyqcGyFeMNwcwlynRnnrJQJHKmKIBvyp8o

This is not normal

**Reply by moi15moi** [reply]
I found this thread: https://stackoverflow.com/questions/71973757/how-to-use-fonttools-to-detect-the-font-type

Here is what I try to compare the two method. Like you can see in the output, there is a pretty big difference when these two should return the same thing.
```py
from matplotlib import font_manager
from fontTools import ttLib

fontsPath = font_manager.findSystemFonts()
numberOfDifference = 0

for path in fontsPath:
    with open(path, 'rb') as fontFile:
        fontType = fontFile.read(4)

    ttLibFont = ttLib.TTFont(path, fontNumber=0)

    isTrueTypeStackOverflow = False
    if fontType.startswith(b'\x00\x01\x00\x00\x00'):
        isTrueTypeStackOverflow = True
    
    isTrueTypeFonttools = False
    if "glyf" in ttLibFont:
        isTrueTypeFonttools = True

    if isTrueTypeFonttools != isTrueTypeStackOverflow:
        numberOfDifference += 1

print("Font: " + str(len(fontsPath)))
print("numberOfDifference: " + str(numberOfDifference))
```

Output:
```
Font: 1048
numberOfDifference: 927
```

**Reply by kenmcd** [reply]
That font is an OpenType file format with TrueType character/glyph outlines.
It comes back as TTF `true` based on the test above, because those TrueType glyphs are in the `glyf` table inside the font.
Usually this type of font has a `.ttf` file extension, but that is not mandatory (but is recommended).
If the font has PostScript outlines, not TrueType outlines, the glyphs are in the `cff` table.
So in that case the above test for the existence of the `glyf` table would be false.

What exactly are you trying to determine?
If a font has TrueType outlines vs. PostScript outlines?
Meaning an OpenType-TT (`.ttf`) font vs. an OpenType-PS (`.otf` usually).
OR
Are you really trying to determine if the font is the old TrueType file format vs. the newer OpenType file format?

EDIT: You posted while I was writing.
That StackOverflow test is to determine the _file format_ as I mentioned above.

**Reply by jenskutilek** [reply]
Depending on your system, searching for font files may include a large number of TrueType collection files, which start with the tag "ttcf" and may contain either TrueType or CFF-based fonts. Your check doesn't include this case respectively counts them as "not TrueType".

**Reply by justvanrossum** [reply]
Your code is buggy: you read four bytes, but then do `startswith` with a five-byte string. That will always return False.

(The stackoverflow code is dodgy: while it does read 5 bytes, I don't know whether the fifth byte will always be zero. It should read four bytes.)

**Reply by justvanrossum** [reply]
Then there are Apple TrueType fonts where the first four bytes are `b"true"`, not `b"\x00\x01\x00\x00"`

**Reply by moi15moi** [reply]
@justvanrossum @jenskutilek  You are both totally right. I just make an wrong test

```py
from matplotlib import font_manager
from fontTools import ttLib

fontsPath = font_manager.findSystemFonts()
numberOfDifference = 0

for path in fontsPath:
    with open(path, 'rb') as fontFile:
        fontType = fontFile.read(5)
    
    if not fontType.startswith(b'ttcf'):

        ttLibFont = ttLib.TTFont(path, fontNumber=0)

        isTrueTypeStackOverflow = False
        if fontType.startswith(b'\x00\x01\x00\x00\x00'):
            isTrueTypeStackOverflow = True
        
        isTrueTypeFonttools = False
        if "glyf" in ttLibFont:
            isTrueTypeFonttools = True

        if isTrueTypeFonttools != isTrueTypeStackOverflow:
            numberOfDifference += 1
            print(path)

print("Font: " + str(len(fontsPath)))
print("numberOfDifference: " + str(numberOfDifference))
```

Output
```
Font: 1048
numberOfDifference: 0
```

So, in brief, everything is good. I did not know that an .otf file is not necessary an opentype font.

My ignorance led me to think looking at the glyphs wasn't a good way to go, but I was wrong.

At least, it confimed that it is a good technic to know if it is an truetype font or an opentype font
